homework_02/task_02.py

```diff
@@ -0,0 +1,25 @@
+"""
+Given an array of size n, find the most common and the least common elements.
+The most common element is the element that appears more than n // 2 times.
+The least common element is the element that appears fewer than others.
+You may assume that the array is non-empty and the most common element
+always exist in the array.
+Example 1:
+Input: [3,2,3]
+Output: 3, 2
+Example 2:
+Input: [2,2,1,1,1,2,2]
+Output: 2, 1
+"""
+from collections import Counter
+from typing import List, Tuple
+
+
+def major_and_minor_elem(inp: List[int]) -> Tuple[int, int]:
+    """
+    Return tuple (most_common_element, least_common_element).
+    """
+    el_occurance = Counter(inp).most_common()
```

Comment:
What is the complexity of this line (Hint: look into source code)?

Could we do better?

Reply:
AFAICT, we have here creating a dictionary and then sorting it by value, which gives an average complexity of O(n * n * log n) = O(n^2 * log n). Sure enough, sorting is a bit overkill when we need just max and min. So, I see how I can reduce the complexity down to O(n^2). Is that it?

Reply:
`dictionary and then sorting it`  create and sort is done one after another so you need to add them O(n) + O(n * log n) -> O(n * log n).

> So, I see how I can reduce the complexity to O(n^2). Is that it?

Almost.  You need O(n)